presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/FunctionCallRewriter.java

```diff
@@ -13,51 +13,140 @@
+import com.facebook.presto.sql.tree.SortItem;
 import com.facebook.presto.sql.tree.SubqueryExpression;
+import com.facebook.presto.sql.tree.Window;
+import com.google.common.base.Splitter;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableMultimap;
+import com.google.common.collect.Multimap;
 
 import java.util.ArrayList;
-import java.util.Arrays;
+import java.util.Collection;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
-import java.util.regex.Matcher;
-import java.util.regex.Pattern;
+import java.util.Optional;
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
-import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.facebook.presto.verifier.framework.VerifierUtil.PARSING_OPTIONS;
+import static com.google.common.base.Preconditions.checkState;
 import static java.util.Objects.requireNonNull;
 
 public class FunctionCallRewriter
-        extends DefaultTreeRewriter<FunctionCallRewriter.Context>
 {
-    public static class FunctionCallSubstitute
+    private static final List<Class<?>> SUPPORTED_ORIGINAL_FUNCTIONS = ImmutableList.of(FunctionCall.class, CurrentTime.class);
+    private static final List<Class<?>> SUPPORTED_SUBSTITUTE_EXPRESSIONS = ImmutableList.of(FunctionCall.class, IfExpression.class, SimpleCaseExpression.class,
+            SearchedCaseExpression.class, Identifier.class, Literal.class, ArrayConstructor.class);
+    private static final String OMIT_IDENTIFIER = "_";
+
+    private final Multimap<String, FunctionCallSubstitute> functionCallSubstituteMap;
+    private final FunctionAndTypeManager functionAndTypeManager;
+
+    private RewriterContext context = new RewriterContext();
+
+    private FunctionCallRewriter(Multimap<String, FunctionCallSubstitute> functionCallSubstituteMap, FunctionAndTypeManager functionAndTypeManager)
     {
-        private final QualifiedName name;
-        private final List<Integer> originalArgumentIndices;
+        this.functionCallSubstituteMap = requireNonNull(functionCallSubstituteMap, "functionCallSubstituteMap is null.");
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+    }
 
-        public FunctionCallSubstitute(QualifiedName name, List<Integer> originalArgumentIndices)
-        {
-            this.name = requireNonNull(name, "name is null");
-            this.originalArgumentIndices = ImmutableList.copyOf(requireNonNull(originalArgumentIndices, "originalArgumentIndices is null"));
+    public static FunctionCallRewriter getInstance(String functionCallSubstitutes, TypeManager typeManager)
+    {
+        Multimap<String, FunctionCallSubstitute> functionCallSubstituteMap = validateAndConstructFunctionCallSubstituteMap(functionCallSubstitutes);
+        if (functionCallSubstituteMap.isEmpty()) {
+            return null;
```

Comment:
nit: Optional (+ flatmap in `QueryRewriter` instead of `map`)

Reply:
Revised